src/guides/resolve-zil-domain.md

```diff
@@ -0,0 +1,515 @@
+
+| constant | description |
+| :--- | :--- |
+| **ZILLIQA_API** | an official Zilliqa endpoint which serves us an entry point to the Zilliqa blockchain |
+| **UD_REGISTRY_CONTRACT_ADDRESS** | Registry address of UD without _0x_ prefix |
+
+We'll discuss the registry contract address later in this guide.
+
+#### Create `index.js`
+
+```javascript
+// index.js
+import hash from 'hash.js';
+
+const ZILLIQA_API = "https://api.zilliqa.com/";
+const UD_REGISTRY_CONTRACT_ADDRESS = "9611c53BE6d1b32058b2747bdeCECed7e1216793";
+```
+
+Next, we need to define and attach a function `resolve` to our HTML button under `id="button"`.
+
+We start writing this function by simply taking our input from the text field and preparing to handle an incorrect domain. We will revisit [error handling](resolve-zil-domain.md#error-handling) later in this guide.
+
+{% hint style="warning" %}
+Any domain that does not end with `.zil` is out of scope for this guide.
+{% endhint }
+
+#### Add .zil resolution
+
+```javascript
+// index.js
+import hash from 'hash.js';
+
+const ZILLIQA_API = "https://api.zilliqa.com/";
+const UD_REGISTRY_CONTRACT_ADDRESS = "9611c53BE6d1b32058b2747bdeCECed7e1216793";
+
+async function resolve() {
+  const userInput = (document.getElementById("input")).value;
+  if (!userInput.endsWith(".zil") {
+    // placeholder for future error handling
+    return;
+  }
+}
+
+document.getElementById("button").addEventListener('click', () => resolve());
+```
+
+## Taking a namehash
+
+Namehashing is an algorithm that tokenizes your domain name in a way that a Zilliqa smart contract can understand.
+
+{% hint style="warning" %}
+It is essential to know the difference between Zilliqa namehashing and [EIP-137](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-137.md#namehash-algorithm). Zilliqa uses **sha256 from SHA-2**, instead of **keccak256** which is more commonly used in Ethereum.
+{% endhint %}
+
+To tokenize our domain we need to split it by the "." character into separate labels, reverse the array, and reduce it to a single hash using a childhash function. Childhash generates a hash of the current label, concatenates it with the parent, and hashes the result as a hex value. For our first parent, we'll take a string of **64 zeros**.
+
+#### Adding a namehash
+
+```javascript
+// index.js
+function namehash(domain) {
+  const parent =
+    '0000000000000000000000000000000000000000000000000000000000000000';
+  return '0x' + [parent]
+    .concat(
+      domain
+      .split('.')
+      .reverse()
+    )
+    .reduce((parent, label) =>
+      childhash(parent, label),
+    );
+}
+
+function childhash(parentHash, label) {
+  parentHash = parentHash.replace(/^0x/, '');
+  const labelHash = sha256(label)
+  return sha256(parentHash + labelHash, "hex");
+}
+
+function sha256(message, inputEnc) {
+  return hash.sha256()
+    .update(message, inputEnc)
+    .digest('hex');
+}
```

Comment:
```suggestion
function namehash(name) {
    if (!name) {
        return new Uint8Array(32);
    }
    const [label, ...remainder] = name.split('.');
    const labelHash = sha256.array(label);
    const remainderHash = namehash(remainder.join('.'));
    return sha256.array(new Uint8Array([...remainderHash, ...labelHash]));
}

function arrayToHex(arr) {
    return '0x' + Array.prototype.map.call(arr, x => ('00' + x.toString(16)).slice(-2)).join('');
}
```

Reply:
Works with `js-sha256`

Reply:
Call it like `arrayToHex(namehash('test.zil'))`

Reply:
This doesn't really produce the same hash as needed.

I ran it with js-sha256 and this is the output I got
```javascript
{
oldHashFunctionWithReduce: "0x5fc604da00f502da70bfbc618088c0ce468ec9d18d05540935ae4118e8f50787",
newHashFunctionWithShaArray: "0xaf2992e5b04145f47ac8ec8b2a55c807c76929834168c3540fd1d9efedd39f1e"
}
```

I think it is because the original namehash function is going between different input encoding. At the start, it uses a normal input as a character array, and only when we hash the concatenated result we are interested in the hex input opposite to the character. The string "0xab"  produces the following hash: `"a50350de8cbfbc43ee04e13e8287390194eaf07db29ec24071f88a03ab5a8541"` while the same input as hex i.e. 0xab produces the following hash: `"0xaf2992e5b04145f47ac8ec8b2a55c807c76929834168c3540fd1d9efedd39f1e"`

Reply:
@JohnnyJumper are you sure that you're testing the same code I've suggested? I can't reproduce the result you're getting.
Both the new and the old versions return `0x5fc604da00f502da70bfbc618088c0ce468ec9d18d05540935ae4118e8f50787` for `brad.zil` for me.
Could you push the changes that you're testing?

Reply:
It was a stupid mistake from my side. I used old namehash function when was testing inside the new one. didn't notice I renamed it for namehash2. My bad